File: bin/splitLocales.js
const path = require('path');
const fs = require('fs');

const I18N_PATH = path.join(__dirname, '..', 'assets', 'i18n');

const contents = fs.readFileSync(path.join(I18N_PATH, 'intl_messages.arb'), { encoding: 'utf8' });
const json = JSON.parse(contents);

const translationStrings = {};
const mainKeys = [];

Object.keys(json).forEach(key => {
  const mainKey = key.split('__')[0];
  const translationKey = key.split('__').splice(1).join('__');

  if (!translationStrings[mainKey]) {
    translationStrings[mainKey] = {};
  }

  translationStrings[mainKey][translationKey] = json[key];

  if (!mainKey.startsWith('@') && !mainKeys.includes(mainKey)) {
    mainKeys.push(mainKey);
  }
});

mainKeys.forEach(mainKey => {
  const output = {};
  const current = translationStrings[mainKey];

  Object.keys(current).forEach(key => {
    output[key] = current[key];
    output[`@${key}`] = translationStrings[`@${mainKey}`][key];
  });

  const str = JSON.stringify(output, null, 2);
  fs.writeFileSync(path.join(I18N_PATH, 'en', `${mainKey}.arb`), str, { encoding: 'utf8' });

  console.log(`Generated ${mainKey}.arb!`);
});

console.log('Done!');

File: bin/split_locales.dart
import 'dart:convert';
import 'dart:io';

void main() {
  final i18nPath = File('${Directory.current.path}/assets/i18n');
  final contents =
      File('${i18nPath.path}/intl_messages.arb').readAsStringSync();
  final Map<String, dynamic> json = jsonDecode(contents);

  var translationStrings = <String, Map>{};
  var mainKeys = <String>[];

  json.keys.forEach((key) {
    var mainKey = key.split('__')[0];
    var translationKey = key.split('__').sublist(1).join('__');

    if (!translationStrings.containsKey(mainKey)) {
      translationStrings[mainKey] = {};
    }

    translationStrings[mainKey][translationKey] = json[key];

    if (!mainKey.startsWith('@') && !mainKeys.contains(mainKey)) {
      mainKeys.add(mainKey);
    }
  });

  const jsonEncoder = JsonEncoder.withIndent('  ');
  mainKeys.forEach((mainKey) {
    var output = {};
    var current = translationStrings[mainKey];

    current.keys.forEach((key) {
      output[key] = current[key];
      output['@$key'] = translationStrings['@$mainKey'][key];
    });

    var str = jsonEncoder.convert(output);
    File('${i18nPath.path}/en/$mainKey.arb').writeAsStringSync(str);
    print('Generated $mainKey.arb.');
  });

  print('Done!');
}
